packages/block-library/src/heading/use-heading-level-validator.js

```diff
@@ -0,0 +1,99 @@
+/**
+ * Returns an array of heading blocks enhanced with the following properties:
+ * path    - An array of blocks that are ancestors of the heading starting from a top-level node.
+ *           Can be an empty array if the heading is a top-level node (is not nested inside another block).
+ * level   - An integer with the heading level.
+ *
+ * @param {?Array} blocks An array of blocks.
+ * @param {?Array} path   An array of blocks that are ancestors of the blocks passed as blocks.
+ *
+ * @return {Array} An array of heading blocks enhanced with the properties described above.
+ */
+function computeOutlineHeadings( blocks = [], path = [] ) {
+	// We don't polyfill native JS [].flatMap yet, so we have to use Lodash.
+	return flatMap( blocks, ( block = {} ) => {
+		if ( block.name === 'core/heading' ) {
+			return {
+				...block,
+				path,
+				level: block.attributes.level,
+			};
+		}
+		return computeOutlineHeadings( block.innerBlocks, [ ...path, block ] );
+	} );
+}
+
+/**
+ * Hook that returns a function to compare a heading level to the preceding one.
+ *
+ * @param {string} currentBlockClientId The client id of the current block.
+ *
+ * @return {Function} Function to compare a given heading level to the preceding
+ *                    one and determine if it is invalid.
+ */
+export default function useHeadingLevelValidator( currentBlockClientId ) {
+	const { headings, isTitleSupported, titleIsNotEmpty } = useSelect(
+		( select ) => {
+			const { getPostType } = select( 'core' );
+			const { getBlocks } = select( 'core/block-editor' );
+			const { getEditedPostAttribute } = select( 'core/editor' );
+			const postType = getPostType( getEditedPostAttribute( 'type' ) );
+
+			return {
+				headings: computeOutlineHeadings( getBlocks() ?? [] ),
+				isTitleSupported: postType?.supports?.title ?? false,
+				titleIsNotEmpty: !! getEditedPostAttribute( 'title' ),
+			};
+		},
+		[]
+	);
+
+	const currentHeadingIndex = headings.findIndex(
+		( { clientId } ) => clientId === currentBlockClientId
+	);
+
+	// Default the assumed previous level to H1.
+	let precedingLevel = 1;
+
+	// If the current block isn't the first Heading block in the content, set
+	// prevLevel to the level of the closest Heading block preceding it.
+	if ( currentHeadingIndex > 0 ) {
+		precedingLevel = headings[ currentHeadingIndex - 1 ].level;
+	}
+
+	const titleNode = document.getElementsByClassName(
+		'editor-post-title__input'
+	)[ 0 ];
+	const hasTitle = isTitleSupported && titleIsNotEmpty && titleNode;
+	const hasMultipleH1 = useMemo(
+		() => headings.filter( ( { level } ) => level === 1 ).length > 1,
+		[ headings ]
+	);
+
+	return ( targetLevel ) => {
+		const levelIsDuplicateH1 = hasMultipleH1 && targetLevel === 1;
+		const levelAndPostTitleMayBothBeH1 =
+			targetLevel === 1 && hasTitle && ! hasMultipleH1;
+		const levelIsTooDeep = targetLevel > precedingLevel + 1;
+		const levelIsInvalid = levelIsDuplicateH1 || levelIsTooDeep;
+		const levelMayBeInvalid =
```

Comment:
I think here we know when `levelIsInvalid` so `levelMayBeInvalid` should be just `levelAndPostTitleMayBothBeH1`, no?

Also you use only `levelMayBeInvalid` exported property in `HeadingLevelDropdown`, but since you know when it is invalid, you could use it to handle some options more specifically there.

It's certainly a big improvement already and not a blocker but maybe a follow-up PR could enhance it a bit more with more specific messages for `invalid` and `maybe invalid`.

Reply:
Yeah, I could go as far as explicitly handling each possible invalidity reason with a specific message, e.g. "The chosen heading level is too deep. Use a lower level or add a lower level heading before this one." But if I do something like that, then I'm not sure the "See the content structure tool for more info" bit should still be there anymore.

I'm not entirely sure how much info should be given in the warning. So for now I'll just keep it simple and as you suggested, explore improvements in a follow-up PR.